src/herder/TransactionQueue.cpp

```diff
@@ -123,91 +58,117 @@ TransactionQueue::tryAdd(TransactionFramePtr tx)
 }
 
 void
-TransactionQueue::remove(std::vector<TransactionFramePtr> const& dropTxs)
+TransactionQueue::removeAndReset(
+    std::vector<TransactionFramePtr> const& dropTxs)
 {
-    for (auto& m : mPendingTransactions)
+    for (auto const& tx : dropTxs)
     {
-        if (m.empty())
+        auto extracted = extract(tx);
+        if (extracted.first != std::end(mPendingTransactions))
         {
-            continue;
-        }
-
-        std::unordered_set<std::shared_ptr<TxMap>> toRecalculate;
-
-        for (auto const& tx : dropTxs)
-        {
-            auto const& acc = tx->getSourceID();
-            auto const& txID = tx->getFullHash();
-            auto i = m.find(acc);
-            if (i != m.end())
-            {
-                auto& txs = i->second->mTransactions;
-                auto j = txs.find(txID);
-                if (j != txs.end())
-                {
-                    txs.erase(j);
-                    if (txs.empty())
-                    {
-                        m.erase(i);
-                    }
-                    else
-                    {
-                        toRecalculate.insert(i->second);
-                    }
-                }
-            }
+            extracted.first->second.mAge = 0;
         }
+    }
+}
 
-        for (auto txm : toRecalculate)
+void
+TransactionQueue::ban(std::vector<TransactionFramePtr> const& dropTxs)
+{
+    auto& bannedFront = mBannedTransactions.front();
+    for (auto const& tx : dropTxs)
+    {
+        auto extracted = extract(tx);
+        for (auto const& tx : extracted.second)
         {
-            txm->recalculate();
+            bannedFront.insert(tx->getFullHash());
         }
     }
 }
 
 bool
-TransactionQueue::contains(TransactionFramePtr tx) const
+TransactionQueue::contains(TransactionFramePtr tx)
+{
+    return find(tx).first != std::end(mPendingTransactions);
+}
+
+TransactionQueue::FindResult
+TransactionQueue::find(TransactionFramePtr const& tx)
 {
-    return std::any_of(std::begin(mPendingTransactions),
-                       std::end(mPendingTransactions),
-                       [&](AccountTxMap const& map) {
-                           auto txMap = map.find(tx->getSourceID());
-                           if (txMap == map.end())
-                           {
-                               return false;
-                           }
-
-                           auto& transactions = txMap->second->mTransactions;
-                           return transactions.find(tx->getFullHash()) !=
-                                  std::end(transactions);
-                       });
+    auto const& acc = tx->getSourceID();
+    auto accIt = mPendingTransactions.find(acc);
+    if (accIt == std::end(mPendingTransactions))
+    {
+        return {std::end(mPendingTransactions), {}};
+    }
+
+    auto& txs = accIt->second.mTransactions;
+    auto distance = tx->getSeqNum() - txs.front()->getSeqNum();
```

Comment:
no need to overcomplicate this code: just use `find_if` to look for the transaction